codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +72,1106 @@
-        if (isOutboundStream(server, streamId)) {
-            if (!(headersFrame instanceof ChannelCarryingHeadersFrame)) {
-                throw new IllegalArgumentException("needs to be wrapped");
+    @Override
+    final void onHttp2Frame(ChannelHandlerContext ctx, Http2Frame frame) {
+        if (frame instanceof Http2StreamFrame) {
+            Http2StreamFrame streamFrame = (Http2StreamFrame) frame;
+            onHttp2StreamFrame(((Http2MultiplexCodecStream) streamFrame.stream()).channel, streamFrame);
+        } else if (frame instanceof Http2GoAwayFrame) {
+            onHttp2GoAwayFrame(ctx, (Http2GoAwayFrame) frame);
+        } else if (frame instanceof Http2SettingsFrame) {
+            Http2Settings settings = ((Http2SettingsFrame) frame).settings();
+            if (settings.initialWindowSize() != null) {
+                initialOutboundStreamWindow = settings.initialWindowSize();
             }
-            childChannel = ((ChannelCarryingHeadersFrame) headersFrame).channel();
-            childChannel.streamId(streamId);
-        } else {
-            ChannelFuture future = bootstrap.connect(streamId);
-            childChannel = (Http2StreamChannel) future.channel();
         }
-
-        Http2StreamChannel existing = childChannels.put(streamId, childChannel);
-        assert existing == null;
     }
 
-    private void onStreamClosed(int streamId) {
-        final Http2StreamChannel childChannel = childChannels.remove(streamId);
+    @Override
+    final void onHttp2StreamClosed(ChannelHandlerContext ctx, Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
         if (childChannel != null) {
-            final EventLoop eventLoop = childChannel.eventLoop();
-            if (eventLoop.inEventLoop()) {
-                onStreamClosed0(childChannel);
-            } else {
-                eventLoop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        onStreamClosed0(childChannel);
-                    }
-                });
-            }
+            childChannel.streamClosed();
         }
     }
 
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
+    @Override
+    final void onHttp2StreamActive(ChannelHandlerContext ctx, Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(stream);
+        ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+        if (future.isDone()) {
+            registerDone(future);
+        } else {
+            future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
+        }
+    }
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)